istioctl/cmd/generate_cert_key.go

```diff
@@ -0,0 +1,225 @@
+	generateKeyAndCertCmd.PersistentFlags().StringVar(&rootCert, "root-cert", "root-cert.pem", "Output root certificate file")
+	return generateKeyAndCertCmd
+}
+
+func generateCert(name string, client kubernetes.Interface, writer io.Writer) error {
+	names := strings.Split(name, ".")
+	workloadHost := "spiffee://cluster.local/" + names[1] + "/" + names[0]
+	// Signer certificate.
+	var signerCert *x509.Certificate
+	// Signer private key.
+	var signerPriv crypto.PrivateKey
+	var err error
+	var signerCertBytes []byte
+	useK8sAsCA := false
+	dep, err := client.AppsV1().Deployments(istioNamespace).Get("istiod", metav1.GetOptions{})
+	if err != nil {
+		return err
+	}
+	for _, env := range dep.Spec.Template.Spec.Containers[0].Env {
+		if strings.Compare(env.Name, "PILOT_CERT_PROVIDER") == 0 &&
+			strings.Compare(env.Value, "kubernetes") == 0 {
+			useK8sAsCA = true
+			break
+		}
+	}
+	if useK8sAsCA {
+		log.Infof("K8s CA is signing certificate...")
+		k8srootcert, err := saveK8sRootCert(client)
+		if err != nil {
+			return err
+		}
+		certChain, keyPEM, _, err := chiron.GenKeyCertK8sCA(client.CertificatesV1beta1().CertificateSigningRequests(),
+			workloadHost, name+".csr.secret", namespace, k8srootcert)
+		if err != nil {
+			return err
+		}
+		rootCrt, err := ioutil.ReadFile(k8srootcert)
+		if err != nil {
+			return err
+		}
+		err = os.Remove(k8srootcert)
+		if err != nil {
+			return err
+		}
+		saveCreds(certChain, keyPEM, rootCrt, writer)
+	} else {
+		log.Infof("Istiod CA is signing certificate...")
+		caSecret, scrtErr := client.CoreV1().Secrets(istioNamespace).Get(ca.CASecret, metav1.GetOptions{})
+		if scrtErr != nil {
+			return scrtErr
+		}
+		signerCertBytes = caSecret.Data[caCertID]
+		signerPrivBytes := caSecret.Data[caPrivateKeyID]
+		signerCert, err = util.ParsePemEncodedCertificate(signerCertBytes)
+		if err != nil {
+			return fmt.Errorf("pem encoded cert parsing failure (%v)", err)
+		}
+		signerPriv, err = util.ParsePemEncodedKey(signerPrivBytes)
+		if err != nil {
+			return fmt.Errorf("pem encoded key parsing failure (%v)", err)
+		}
+		opts := util.CertOptions{
+			Host:       workloadHost,
+			NotBefore:  getNotBefore(),
+			TTL:        validFor,
+			SignerCert: signerCert,
+			SignerPriv: signerPriv,
+			Org:        org,
+			IsCA:       false,
+			IsClient:   true,
+			RSAKeySize: keySize,
+		}
+		certPem, privPem, err := util.GenCertKeyFromOptions(opts)
+		if err != nil {
+			log.Fatalf("Failed to generate certificate: %v\n", err)
+		}
+		saveCreds(certPem, privPem, signerCertBytes, writer)
+	}
+	return nil
+}
+
+func saveK8sRootCert(client kubernetes.Interface) (string, error) {
+	secrets, err := client.CoreV1().Secrets(istioNamespace).List(metav1.ListOptions{})
+	if err != nil {
+		return "", err
+	}
+	for _, secret := range secrets.Items {
+		if strings.HasPrefix(secret.Name, "default-token") {
+			f, err := os.Create(path.Join(os.TempDir(), "k8s-ca.crt"))
+			if err != nil {
+				return "", err
```

Comment:
Can you prefix these errors so the user know what was being attempted when it failed?